### ISSUE TITLE
No L2 Construct for EKS

### ISSUE THREAD

**Issue by IPyandy** [awslabs/aws-cdk#990]
Right now in order to create a workable EKS Cluster with worker nodes, one has to fall back to L1 cloudformation construct. This makes it difficult to use CDK for purposes of setting up any type of managed kubernetes cluster.

The user must:

- Use VpcNetwork to create the network
- Use cloudformation construct to deploy the cluster
- Manually create autoscaling groups
- Properly tag autoscaling group instances
- Create user data to match the bootstrap script
- Create a role
- Attach managed policies to the role
- Apply securityGroup rules by following the [Getting Started Guide](https://docs.aws.amazon.com/eks/latest/userguide/getting-started.html)

As the nature of EKS clusters and more so worker nodes are probably added and removed from existence quite often, especially during dev and testing.

## Design Case

The user should be able to create a cluster, by passing in some properties. The only required properties in the `cloudformation` construct are the `VPC` and the `VpcPlacement`, such as.

```ts
const cluster = new eks.Cluster(clusterStack, "Cluster", {
  vpc: clusterVpc,
  vpcPlacement: {
    subnetsToUse: ec2.SubnetType.Public
  }
});
```

Which would then create

- A security group to attach to the cluster
- The correct role for the cluster
    - with provided managed policies by AWS
- Make certain properties public for access during node creation
    - the cluster name
    - the cluster endpoint
    - others for future reference
- Be able to `import` and `export` available for future multi-stack deployments

The user can then add worker nodes on demand of any of the supported classes and sizes. The only required inputs would be the `VPC` to and which `Cluster (control plane)` to use for the nodes. Optional properties can be added to scale the autoscaling group, provide `ssh keypair` if SSH access to the nodes is required.

These options can be added in future releases to make on-par with `autoscaling group` options as well, but provide an abstraction to those.

The nodes should automatically

- Tag the instances with required EKS tags
- Create the node role and instance profile
    - with correct policies
- Automatically add the userdata for startup script and cluster discovery
- Add recommended security group rules to cluster and nodes based on the [getting started documentation](https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html)
- Track an array of node autoscaling groups created

### Create Worker Nodes (default values)

```ts
const grp1 = new eks.Nodes(clusterStack, "NodeGroup1", {
  vpc: clusterVpc,
  cluster
});
```

### Create Worker Nodes (custom values)

```ts
const grp2 = new eks.Nodes(clusterStack, "NodeGroup2", {
  vpc: clusterVpc,
  cluster,
  nodeClass: ec2.InstanceClass.T2,
  nodeSize: ec2.InstanceSize.Medium,
  nodeType: eks.NodeType.Normal,
  minNodes: 2,
  maxNodes: 4,
  sshKeyName: "aws-dev-key"
});
```

I'll attempt to create a PR for this, feedback is welcome, will be my first contribution here if accepted.

Otherwise, I'll learn from the review :)

**Comment by PaulMaddox** [comment]
I’d be happy to contribute / help review this

**Comment by ijcd** [comment]
Some quick feedback. This structure is missing info on T3 insances, which leads to `max` being `undefined` in the instance user-data. Seems to prevent nodes from registering.

```
export const maxPods = Object.freeze(
```

```
#!/bin/bash
set -o xtrace
/etc/eks/bootstrap.sh stargatedemoclusterClusterA-oChms63INSG8 --use-max-pods undefined
```

Probably need some typed checking on the hash lookup to prevent misses like this from sneaking through.

**Comment by ijcd** [comment]
I also ended up adding a `securityGroups` option to `IClusterProps` so we could use some imported security groups.

```
  securityGroups?: [ec2.SecurityGroupRef]
```

```
    const securityGroupIds = new Array(this.securityGroupId)
    if (props.securityGroups) {
      for (var sg of props.securityGroups) {
        securityGroupIds.push(sg.securityGroupId)
      }
    }
```

**Comment by eladb** [comment]
We released an EKS library in 0.25.0
